### ISSUE TITLE
Wildcards in map and list patterns lead to "unnecessary" operations

### ISSUE THREAD

**Issue by stereotype441** [dart-lang/language#2671]
As currently specified, matching an expression `e` against a pattern like `{'foo': _}` causes three map operations to be performed:
- check if `e.length` is equal to 1
- check if `e.containsKey('foo')` returns `true`
- evaluate `e['foo']` and discard the result

For a standard implementation of `Map`, the third bullet has no effect.

Similarly, matching an expression `e` against a pattern like `[_]` causes two list operations to be performed:
- check if `e.length` is equal to 1
- evaluate `e[0]` and discard the result

For a standard implementation of `List`, the second bullet has no effect.

Since the user may have their own implementations of `Map` and `List`, these "unnecessary" operations could actually have observable effects in contrived programs.  Because of that, it may be difficult for the compiler to optimize out the unnecessary operations even in programs where they truly have no effect.

We could fix this by specifying that when a map or list pattern contains a wildcard pattern, the call to `operator[]` is not performed.  Should we?

**Comment by stereotype441** [comment]
CC @munificent @leafpetersen @jakemac53 @lrhn @mit-mit @natebosch @eernstg @kallentu

**Comment by eernstg** [comment]
I'd be happy about the specification where those invocations of `operator []` are omitted.

**Comment by lrhn** [comment]
We've already defined member accesses as lazy variables, to ensure they are read only once. 

Would it make sense to change the pattern semantics so that the _variable_ is provided, instead of the value, so that if the pattern never actually references the variable's value, it won't be read and initialized?

There aren't that many patterns which don't look at the value at all. The `_` or its `var _` or `final _` variants are probably the only ones. Just special casing that pattern could also do it. 

Basically an **usesValue**(pattern) predicate on patterns, and then before each member access which is just there to provide a value for a pattern, we check whether *usesValue**(pattern), and if not, we don't do the member access.

(It's very likely that it'll only ever  be used for list patterns like `["kindMarker", _, var value2]`, but I do think it's a good choice to not do that index operation.)

**Comment by munificent** [comment]
I did just change the runtime specification for map patterns to say:

```

        A compiler is free to call `v[k]` and `containsKey()` in either order,
        or to elide calling one or both if it determines that doing so will
        produce the same result. It may assume that the map adheres to the
        following protocol:

        *   If `containsKey(k)` returns `false` for some key, then `v[k]` will
            return `null`.

        *   If `containsKey(k)` returns `true` for some key, then `v[k]` returns
            an instance of the map's value type.

        *In particular, if the map's value type is non-nullable, then when
        `v[k]` returns `null`, the compiler can assume that the key is absent
        and `containsKey(k)` would return `false` too.*
```

It doesn't specifically call out cases where the subpattern doesn't need the value. Perhaps "same result" is a little too vague, but my intent was to allow cases like this too.

I could do something similar for list patterns.

**Comment by stereotype441** [comment]
@munificent 
Oh, I didn't realize that was there.  Hmm.

I can't decide what I think about leaving these "optimizations" up to the implementation.  On the one hand, I like that it makes less specification work for us to do, and gives the CFE team more freedom to be smart and do the right thing.  On the other hand, I worry about someone coming to rely on a particular implementation behaviour and then running into trouble if we later change it.

For example, imagine someone has a code that matches a `Map<String, int>` against the pattern `{'foo': _}`.  The spec text permits the implementation to implement this as either `x.containsKey('foo')` or `x['foo'] != null`.  Now imagine they decide to unit test their code using Mockito.  They're going to create an expectation that either `containsKey` is called or an expectation that `operator[]` is called, depending what the implementation does.  Now, if they upgrade to a new version of Dart that implements the pattern differently, their test is going to fail.

**Comment by leafpetersen** [comment]
> I did just change the runtime specification for map patterns to say:

I have some concerns about this doing this.  This is *really* doubling down on allowing undefined behavior.  I know we crossed that rubicon with making identity on records undefined, but this feels qualitatively different to me.  A question that it immediately brings to mind is, why do this here, and not in any of the many other places where the language is defined against a protocol?  For example, we could say that a `for in` loop is not required to call `.current` if the value is not needed, but I don't think we do.  Similarly for spreads, etc.  So it feels a bit odd to call this one place out where we're going to rely on an informal protocol specification to define the semantics of the language.

**Comment by munificent** [comment]
I am definitely not attached to this approach. I personally prefer syntax that desugars to simple stable uses of protocols. But I also know that I tend to err on the side of "performance doesn't that matter that much most of the time and you can always not use the sugar if you don't like the perf" and have gotten pushback on it.

I'll re-open #2634 and we can discuss it there and here. @eernstg and @lrhn seemed to feel strongly that the optimization benefits were worth it.

**Comment by eernstg** [comment]
> @eernstg and @lrhn seemed to feel strongly that the optimization benefits were worth it.

Not really&mdash;my motivation for allowing the implementation specific behavior was based on a general preference for specifying the semantics of the language to a certain degree, and then leaving some space open for optimizations. In particular, I tend to think that developers "should not rely on" side effects of calling `operator []` in the situations described here. It seems overly strict to specify that these optimizations are forbidden, but also overly constraining to specify that a very precise set of optimizations _must_ be implemented.

I do hear @leafpetersen's and others' worries about the predictability/debuggability of programs with this kind of implementation specific wiggle room. Also, I do recognize that we made the choice to _avoid_ the wiggle room in several situations. E.g., we specify a precise evaluation order for actual arguments in function invocations (in contrast to a language like C, where it is 'undefined behavior' to rely on any specific evaluation order, even though probably every single C program _does_ rely on it to some extent). However, I do think it's a deep tar pit to explicitly make the evaluation order of actual arguments undefined; the dangers associated with the missing `operator []` invocation seem much smaller to me.

I really don't know where to draw the line.

**Comment by stereotype441** [comment]
I don't know if I could come up with a good general principle for where to draw the line, but as I've been thinking about it, I feel more and more convinced about what I would like to do for this case in specific: just go ahead and specify precisely what we want to happen in the desugaring (and write language tests to lock down the implementation).

Personally I would probably say something like this (in the "Matching (refuting and destructuring)" section):

> *   **Map**:
>     1. ...
>     2. ...
>     3. ...
>     4.  Otherwise, for each (non-rest) entry in `p`, in source order:
>
>         1.  Evaluate the key `expression` to `k`.
>         2.  If the entry's value subpattern is a wildcard pattern and `V` is a subtype of its required pattern (*in other words, we only need to test whether the key is contained in the map; we don't need to examine the value*):
>             1.  Call `containsKey(k)` on the value. If this returns `false`, the map does not match.
>         3.  Otherwise, if `Null` is not a subtype of `V` (*in other words, `V` is non-nullable, so if `operator[]` returns `null`, we know that the key is not present*):
>             1.  Evaluate `v[k]` and check if the resulting value is `null`. If it is, the map does not match.
>             2.  Otherwise, match the resulting value against this entry's value subpattern. If it does not match, the map does not match.
>         4.  Otherwise (*the general case, where we must call both `containsKey` and `operator[]` to determine whether the pattern matches*):
>             1.  Call `containsKey(k)` on the value. If this returns `false`, the map does not match.
>             2.  Otherwise, evaluate `v[k]` and match the resulting value against this entry's value subpattern. If it does not match, the map does not match.
>     5.  The match succeeds if all entry subpatterns match.

As a side note, even with normally-behaving maps, one of these "optimizations" is not perfectly semantics-preserving: in mixed mode code, even if `V` is non-nullable, it's possible for `operator[]` to return `null` even if the key is present (because an unmigrated library could have inserted a `null` value into the map).  But I don't think that's a problem.

There's one other optimization we could consider (I can't decide if it's worth it): in the general case, rather than always evaluating both `containsKey(k)` and `v[k]`, we could evaluate `v[k]` first, and then only call `contains(k)` if the result is `null`.  On the one hand, it's almost certainly faster (since the cost of the null check and branch is small compared to the cost of a second map lookup).  On the other hand, it's not the way most people would hand-desugar a map pattern, so it might lead to user surprise when mocks or custom map implementations are in use.

**Comment by lrhn** [comment]
I really prefer to not overspecify. I've seen what that does to implementations. 
JavaScript have had some issues, where you really can't both do the best thing and satisfy the specification, because the specification decided which order to do things in, just because it had to specify *something*.

On the other hand, I've also seen, at least as often, that unspecified behavior gets locked in anyway, by users depending on whatever the current implementation does. If you can't change it anyway, you might as well specify it. At least it'll be an explicit and deliberate breaking change if you ever do decide that changing behavior is worth it.

Rock. Specification. Hard place.

When you specify unspecified behavior, you pretty much need to implement *randomized* behavior, just to keep people from assuming. Which has a cost too, even if you only randomize in dev builds, and use the (currently) most performant approach in release builds.

"Tradeoffs are hard, let's go shopping."

I'd be fairly happy with a choice of:
```dart
late final bool containsKeyX = valueX != null || map.containsKey(keyX);
late final V valueX = map[keyX];
```
where we allow the compiler to omit the `map.containsKey(keyX)` if a map value cannot be `null`.

Most maps won't map keys to `null`, because it's not particularly useful, so it should usually be a win.
It should be *fast enough*. You will do value lookups, even when you don't need the value, but not needing the value isn't going to be common, and it happening for a `null` value isn't going to be common either.

This all assumes that `operator[]` and `containsKey` are comparable in efficiency. That is almost always going to be the case. If `containsKey` is significantly faster, it's probably a sucky map implementation.
(But I'm sure they exist.)

**Comment by stereotype441** [comment]
I did some thinking this morning about the relationship between map patterns, flow analysis, and soundness, and I filed https://github.com/dart-lang/language/issues/2685.  The TL/DR is that as things are currently defined, perfectly reasonable map patterns like `{'foo': var x}` are unsound, because they bind `x` to the result of calling `operator[]` without null checking it, but assign `x` a static type of `V` rather than `V?`.  This unsoundness is not limited to mixed mode programs--it happens in fully null-safe programs too.  I think we need to fix the soundness problem even if we decide not to fully specify the runtime semantics of map pattern matching.

Taking @lrhn's suggstion as a starting point:

> ```dart
> late final bool containsKeyX = valueX != null || map.containsKey(keyX);
> late final V valueX = map[keyX];
> ```
> 
> where we allow the compiler to omit the `map.containsKey(keyX)` if a map value cannot be `null`.

This still has the same soundness problem, because a class that returns `true` from `containsKey`, and returns `null` from `operator[]`, will result in `containsKeyX = true` and `valueX = null`, even if `V` is a non-nullable type.

We could fix this by changing it to:

```dart
late final bool containsKeyX = lookupX != null || map.containsKey(keyX);
late final V? lookupX = map[keyX];
late final V valueX = lookupX as V;
```

With the understanding that the compiler is allowed to omit `as V` if `V` is nullable (an understanding that hardly needs to be stated, since this is a semantics-preserving optimization).

But at this point I'm not convinced we've really gained much benefit from being loose with the specification, because the only wiggle room we're allowing is for the implementation to decide whether or not to omit `|| map.containsKey(keyX)` if `V` is non-nullable, an optimization that we almost certainly want to perform whenever we can.

**Comment by lrhn** [comment]
We make some assumptions, which is always dangerous for soundness.
* If `map[key]` is non-null, then `map.containsKey(key)` is true.
* If `map.constainsKey(key)` is true, then `map[key]` is `V` (and not `null` if `V` is not nullable).

The first assumption is sound (doesn't break type safety). Pattern matching a non-compliant implementation might not work as expected, but that's what you get for not complying with the `Map` protocol.

The second assumption is *not* safe. If we read `map.containsKey(key)` first and then blindly use the value of `map[key]` as `V`, we break type soundness, if `V? operator[]` returns `null` anyway.

All that to get to the same conclusion: We need the lookup result separate from the value.

Then it's just about how we best optimize the number of lookups.
```dart
late final V? lookupX = map[keyX];
late final bool containsKeyX = 
      lookupX != null && (valueX = lookupX) != null                        // Assume we can promote lookupX
   || map.containsKey(keyX) && (valueX = null as V) == null;
late final V valueX;
```
This eagerly initializes `valueX` if `containsKeyX` is true, and not otherwise, but we should never read `valueX` without first checking `containstKeyX` anyway. (We already checked that `lookupX` was non-`null` in `containsKeyX`, so we might as well use it immediately.)

**Comment by munificent** [comment]
I'd like to get this issue resolved and closed. The current status of map pattern matching is:

1. It's now sound with respect to misbehaving maps and `null`, so we don't have to worry about it here.

2. It doesn't leave the behavior around calling `containsKey()` and `v[key]` up to the compiler, but it does [carefully specify which operations to call](https://github.com/dart-lang/language/blob/master/accepted/future-releases/0546-patterns/feature-specification.md#matching-refuting-and-destructuring) such that it should be as efficient as possible based on whether the value type is known to be nullable or not.

So the comments here discussing those are I believe resolved. We can get back to the main point of this issue which is **whether we should elide accessing map and/or list elements when the subpattern it will be passed to is a wildcard.**

My take is:

### Wildcards in maps

We should *not* elide the access for map patterns. Even when the subpattern doesn't use the value, the map pattern still needs to see if the key is present. That's why you might write a map pattern entry with a wildcard in the first place. (And, unlike list patterns, it's the *only* reason to write one.)

According to the current spec (which is different from when this issue was first files), when you write:

```dart
test(Map<String, int> things) {
  if (things case {'a': _}) {
    ...
  }
}
```

Then after checking the length of `things`, the pattern calls `things['a']`. If that returns a non-null value, it knows the key is present and doesn't call `containsKey()`. In cases where the value type is nullable, it may need to call both the index operator and `containsKey()` to tell if the key is present. But in cases where the value type is known to be non-nullable, it only calls `v[key]`, so there isn't anything we can optimize out.

I don't think it's a clear win to say that *if* the map has a potentially nullable value type and the value subpattern is a wildcard, then instead of calling `v[key]` and then calling `containsKey()` if the result is `null`, we only (but unconditionally) call `containsKey()` instead.

It feels like we're just doing speculative optimization. I think we should just leave maps alone. We've poked at the runtime semantics enough.

### Wildcards in objects

While we're at it, [wildcards in object patterns have also been raised](https://github.com/dart-lang/sdk/issues/51029
). For those, I think we *should* call the getter. It might have a side effect and the user did write the pattern, presumably for some reason.

### Wildcards in lists

For list patterns, I think it's more reasonable (but probably only marginally valuable) to elide the index access. Unlike map entries, you might write a wildcard list element pattern when you don't care about that element at all. You just want to shift the index of later subpatterns:

```dart
var [_, _, _, d] = [1, 2, 3, 4];
```

This pattern doesn't care about the first three elements of the list at all. The `_, _, _, ` is just there to change the index of the element we access for `d`.

At the same time, this would make lists inconsistent with what I think we should do for maps and objects. So I'm inclined to not do the optimization here either.

This keeps the behavior simple and predicatable. If you're really in a weird situation where the runtime cost of accessing a list element is a problem... you can always not use a pattern.

Does that sound reasonable?

**Comment by eernstg** [comment]
Sounds very reasonable! I think this allows us to outline the rules concisely, which is an important benefit:

> A wildcard pattern causes the corresponding getter/operator[] to be invoked, and the value is then discarded. When the value is null and the pattern is a map pattern whose value type is potentially nullable, `containsKey` is called to distinguish a missing key from a key with the value null.
> *This confirms the existence of the denoted property, and it causes the side effects to occur, if any.*

**Comment by stereotype441** [comment]
@munificent said:
> ... a lot of reasonable stuff ...
> 
> Does that sound reasonable?

Yes!  Thanks for cicling back to this issue and bringing such clarity to it.  I really like what we've settled on here.

**Comment by lrhn** [comment]
I'd prefer if list patterns do not do anything for wildcard element patterns.

It's probably rare, but possible, to have a list which performs computation on access. One classical example could have been the captures of a `RegExp` match, if we had exposed them as a list (which I'm suggesting to do if we rework the class). The underlying implementation doesn't allocate anything, it just finds the start and end of a capture.
The match would the lazily allocate the captured string or sub-match object if you access it, and won't if you don't.
(The V8 `RegExp` result probably still works like that.)

So, if I do `switch (m.captures) { case [_, _, var digits?] => int.parse(digits); ...}`, I'd be annoyed if it caused allocation of the two first entries.

We can say that it's only the `_` pattern which does nothing, if you write `var _` or any other pattern structure, we treat it as a normal pattern, but I'm not sure there's any advantage in that.


The rest looks fine.

**Comment by eernstg** [comment]
@lrhn wrote:
> It's probably rare, but possible, to have a list which performs computation on access

We might then have side effects, too, which could actually serve as an argument why we should _not_ silently skip the lookup of the first two elements, we should stick to the simple rule that an `_` gives rise to a lookup, and the result is then discarded.

However, we could just improve on the ability to use a pattern that doesn't seem to perform those lookups:

```dart
switch (m.captures) { case List([2]: var digits?] => int.parse(digits); ...}
```

With this kind of pattern it's obvious that no other elements than the one at index 2 is looked up. Also, this will still be usable when we want element `[200]`.

We could generalize this idea in several ways (e.g., we could call methods, passing an all-constant list of actual arguments, or we could call other operators, etc), but that's just an open space that we can choose to explore later on.

With that in mind, I'd still prefer to have a simple and consistent rule about the semantics, and then recommend that developers who _must_ avoid those lookups to use operator `[]` rather than a pattern, at least until `-later`.

**Comment by lrhn** [comment]
Yup, allowing any selector(-chain) in an object pattern is on my "patterns-later" list.

It avoids people doing obnoxious workarounds like:
```dart
extension <T> on List<T> {
   T get g200 => this[200];
}

....
   case List<int>(length: >200, g200: var x): ... print(x);
```

Nothing goods come from making `_` read the element of a list. 
Using a `_` place-holder to reach a specific position is the most obvious and obviously idiomatic way to match only a later element: Show `switch (list) { case [_, _, var x, ...] => x, _ => null}` to *anybody* and they'll say it's equivalent `var x = list.length > 3 ? list[x] : null;`.

I'd bet that even if you tell them that it isn't, they can't possibly imagine how it could differ. Actually reading the first two elements is so far from what anyone would expect, that only someone coming from the specification side can imagine it.

(And it's different from `SomeType(x: _)` and `{"x": _}` because both of those actually mention the thing to read, and they could just as easily *not* do that. The list pattern cannot omit mentioning the position.


If we insist that `_` will read, I propose we introduce *elided* element patterns in list patterns, so you can write:
```dart
  case [,, var x]: ...
```
where the empty/elided patterns do not read anything, where `[_, _, var x]` would then actually read the value and do nothing with it. (Which does enforce that the getter doesn't throw, which a `list.cast` result might.)

That'd be fine too, but I don't want `_` to read the element and *not* have another way to avoid reading it.

**Comment by munificent** [comment]
OK, I'm persuaded to elide the element accesses for lists: https://github.com/dart-lang/language/pull/2797.
